toolchain/semantics/semantics_handle_index.cpp

```diff
@@ -15,79 +15,89 @@ auto HandleIndexExpressionStart(Context& /*context*/,
+                                        SemIR::Node operand_node,
+                                        SemIR::Node index_node, int size)
+    -> const llvm::APInt* {
+  const auto& index_val = context.semantics_ir().GetIntegerLiteral(
+      index_node.GetAsIntegerLiteral());
+  if (index_val.uge(size)) {
+    CARBON_DIAGNOSTIC(IndexOutOfBounds, Error,
+                      "Index `{0}` is past the end of `{1}`.", llvm::APSInt,
+                      std::string);
+    context.emitter().Emit(
+        parse_node, IndexOutOfBounds,
+        llvm::APSInt(index_val, /*isUnsigned=*/true),
+        context.semantics_ir().StringifyType(operand_node.type_id()));
+    return nullptr;
+  }
+  return &index_val;
+}
+
 auto HandleIndexExpression(Context& context, ParseTree::Node parse_node)
     -> bool {
-  CARBON_DIAGNOSTIC(OutOfBoundsAccess, Error,
-                    "Index `{0}` is past the end of `{1}`.", llvm::APSInt,
-                    std::string);
-
   auto index_node_id = context.node_stack().PopExpression();
   auto index_node = context.semantics_ir().GetNode(index_node_id);
-  auto name_node_id = context.node_stack().PopExpression();
-  name_node_id = context.MaterializeIfInitializing(name_node_id);
-  auto name_node = context.semantics_ir().GetNode(name_node_id);
-  auto name_type_id =
-      context.semantics_ir().GetTypeAllowBuiltinTypes(name_node.type_id());
-  auto name_type_node = context.semantics_ir().GetNode(name_type_id);
+  auto operand_node_id = context.node_stack().PopExpression();
+  operand_node_id = context.MaterializeIfInitializing(operand_node_id);
+  auto operand_node = context.semantics_ir().GetNode(operand_node_id);
+  auto operand_type_id = operand_node.type_id();
+  auto operand_type_node = context.semantics_ir().GetNode(
+      context.semantics_ir().GetTypeAllowBuiltinTypes(operand_type_id));
 
-  if (name_type_node.kind() == SemIR::NodeKind::ArrayType) {
-    auto [bound_id, type_id] = name_type_node.GetAsArrayType();
-    if (index_node.kind() == SemIR::NodeKind::IntegerLiteral) {
-      const auto& index_val = context.semantics_ir().GetIntegerLiteral(
-          index_node.GetAsIntegerLiteral());
-      if (index_val.uge(context.semantics_ir().GetArrayBoundValue(bound_id))) {
-        context.emitter().Emit(
-            parse_node, OutOfBoundsAccess,
-            llvm::APSInt(index_val, /*isUnsigned=*/true),
-            context.semantics_ir().StringifyType(name_node.type_id()));
+  switch (operand_type_node.kind()) {
+    case SemIR::NodeKind::ArrayType: {
+      auto [bound_id, element_type_id] = operand_type_node.GetAsArrayType();
+      // We can check whether integers are in-bounds, although it doesn't affect
+      // the IR for an array.
+      if (index_node.kind() == SemIR::NodeKind::IntegerLiteral) {
+        ValidateIntegerLiteralBound(
+            context, parse_node, operand_node, index_node,
+            context.semantics_ir().GetArrayBoundValue(bound_id));
+      }
```

Comment:
Curious why not produce an error in the SemIR for this?

One thing I could imagine wanting is to add some way of modelling errors that doesn't *remove* the SemIR information so that we can find the `-1` index operation. If that's the direction you're imagining, maybe leave a TODO?

Reply:
The diagnostic is sufficient to mark the IR as invalid.

It sounds though like you really want the IR to specifically include an error, so I've changed code to add logic that replaces the index node id with an error.